rfd/00xx-kubernetes-access-cluster-discovery.md

```diff
@@ -0,0 +1,362 @@
+
+### Scope
+
+This RFD focuses only on AWS EKS clusters. Similar ideas will be explored in the future for GCP's GKE and Azure's AKS.
+
+## Details
+
+### AWS EKS discovery and IAM
+
+AWS API has a method that allows listing every EKS cluster by calling [`eks:ListClusters`][listclusters] endpoint. This endpoint returns all EKS cluster names to which the user has access. The response has no other details besides the cluster name. So, for Teleport to extract the cluster details such as CA, API endpoint, and labels, it has to make an extra call, per cluster, to [`eks:DescribeCluster`][descclusters]. 
+
+The necessary IAM permissions required for calling those two methods are:
+
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "eks:DescribeCluster",
+                "eks:ListClusters"
+            ],
+            "Resource": "*" # can be limited
+        }
+    ]
+}         
+```
+
+Besides the previous IAM policies required for the discovery service, the discovery service must also have the `sts:GetCallerIdentity` IAM permission to generate an access token to Kubernetes API.
+
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "sts:GetCallerIdentity"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+With cluster details such as API endpoint and CA, Teleport creates an access token to gain access to the cluster. The access is given by a token generated by [AWS IAM authenticator][awsiamauthenticator] project. This project generates a short-lived user access token by mapping the user IAM username/role into [Kubernetes RBAC credentials][iamroleatch] ([eksctl IAM Mappings][iameksctl]).
+
+AWS IAM does not allow setting access permissions to any Kubernetes cluster. This means that it is not possible through the AWS API to guarantee access to one or several Kubernetes clusters. This is achieved by an extra project, [AWS IAM authenticator][awsiamauthenticator], that is installed, by default, in the EKS control plane which translates the IAM roles and users into Kubernetes roles or users. [AWS IAM authenticator][awsiamauthenticator] component receives a request when any user generates an access token for a cluster and, it searches in its database for any match between IAM user or role and the Kubernetes role. If the match is found, it returns a short-lived access token for the impersonated Kubernetes user or role, otherwise, it returns a token for `system:anonymous` group which has no access to resources.
+
+The [AWS IAM authenticator][awsiamauthenticator] database is a simple `configmap/aws-auth` stored in `kube-system` namespace. This configmap cannot be changed unless Teleport has write permissions to Kubernetes Configmaps, which can only happen if the Teleport role is already in the configmap.
```

Comment:
Does this config map always exist in an EKS cluster?

Reply:
For now yes. 

For freshly installed clusters, the `aws-auth` configmap is already populated with the mapping between the node's IAM role and Kubernetes roles otherwise nodes cannot join the cluster.

AWS IAM authenticator also has support for CRDs (alpha status and not enabled for EKS):
```yaml
---
apiVersion: iamauthenticator.k8s.aws/v1alpha1
kind: IAMIdentityMapping
metadata:
  name: kubernetes-admin
spec:
  # Arn of the User or Role to be allowed to authenticate
  arn: arn:aws:iam::XXXXXXXXXXXX:user/KubernetesAdmin
  # Username that Kubernetes will see the user as, this is useful for setting
  # up allowed specific permissions for different users
  username: kubernetes-admin
  # Groups to be attached to your users/roles. For example `system:masters` to
  # create cluster admin, or `system:nodes`, `system:bootstrappers` for nodes to
  # access the API server.
  groups:
  - system:masters

```